language-hask-source/src/Language/Hask/Parsed/AST.hs

```diff
@@ -0,0 +1,575 @@
+  , importHiding    :: Bool             -- ^ Presence of @hiding@ keyword.
+  , importList      :: [Import]
+  } deriving stock Show
+
+-- | Parsed import.
+data Import
+  = VariableImport VariableName
+    -- ^ Import of a simple variable.
+  | TypeImport ConstructorName (Maybe TypeImportList)
+    -- ^ Import of a type.
+    --
+    -- > $1
+    --
+    -- or
+    --
+    -- > $1 ($2)
+    --
+    -- See [Parsed type import/export lists](#ParsedTypeImportLists).
+  deriving stock Show
+
+-- | Parsed import/export list of a type. You may be interested in
+-- [Parsed type import/export lists](#ParsedTypeImportLists).
+data TypeImportList
+  = TypeImportAll
+    -- ^ Import/export of all members in scope.
+    --
+    -- > (..)
+  | TypeImportList [Either VariableName ConstructorName]
+    -- ^ Import/export of explicitly named members.
+    --
+    -- > ($1)
+  deriving stock Show
+
+-- $ParsedTypeImportLists
+-- == Parsed type import/export lists #ParsedTypeImportLists#
+-- 'TypeImportList's contain both class and datatype imports/exports, because
+-- we can't yet distinguish those properly - that's up to the renamer.
+
+-------------------------------------------------------------------------------
+-- | Parsed top-level declaration.
+data Declaration
+  = TypeSynonym Type Type
+    -- ^ Type synonym declaration.
+    --
+    -- > type $1 = $2
+  | Datatype Type DatatypeBody (Maybe Deriving)
+    -- ^ Datatype declaration.
+    --
+    -- > data $1 = $_ | ... $3
+    --
+    -- or
+    --
+    -- > newtype $1 = $2 $3
+    --
+    -- See [Types in declaration heads](#TypesInDeclarationHeads).
+  | ClassDeclaration Class
+  | InstanceDeclaration Instance
+  | Defaults [Type]
+    -- ^ "default" declaration.
+    --
+    -- > default ($1)
+  | Binding Binding
+  deriving stock Show
+
+-- | Parsed datatype body.
+data DatatypeBody = Newtype NewtypeConstructor | Data [DataConstructor]
+  deriving stock Show
+
+-- | Parsed deriving clause.
+--
+-- > deriving $_
+--
+-- or
+--
+-- > deriving ($_, ...)
+type Deriving = Either (Qualified ConstructorName) [Qualified ConstructorName]
+
+-- | Parsed "newtype" constructor.
+--
+-- > $newtypeConstructorName { $newtypeConstructorLabel :: $newtypeConstructorField }
+data NewtypeConstructor = NewtypeConstructor{
+    newtypeConstructorName  :: ConstructorName
+  , newtypeConstructorLabel :: Maybe VariableName
+  , newtypeConstructorField :: Type
+  } deriving stock Show
+
+-- | Parsed "data" constructor.
+--
+-- > $dataConstructorName $dataConstructorFields
+data DataConstructor = DataConstructor{
+    dataConstructorName   :: ConstructorName
+  , dataConstructorFields :: Either [Type] [(VariableName, Strictness, Type)]
+    -- ^ "data" constructor fields, possibly named.
+    --
+    -- > $_ ...
+    --
+    -- or
+    --
+    -- > { $_ :: $_, ... }
+  } deriving stock Show
+
+-- | Parsed presence of strictness flag ("bang" @!@).
+type Strictness = Bool
```

Comment:
If we leave this as a `Bool` could we add something like the following synonyms:

```haskell
pattern Strict :: Strictness
pattern Lazy   :: Strictness
pattern Strict = True
pattern Lazy   = False
```

Reply:
Sounds good, will do.